### PR TITLE
[contracts] Implement transparent hashing for contract storage

### LINKED ISSUE #11029: Implement transparent hashing for contract storage
Right now all the storage functions used by contracts take a fixed 32 byte key as input. The contracts pallet hashes this key with `blake2_256`.

This completely removes the possibility to iterate over `(key, value)` pairs. The fixed key size also forces contracts to either hash or pad their keys which adds additional complexity.

We should do the following:
* Add a new version for all storage primitives that take a variable sized key.
* Hash the key with `blake2_128_concat` instead of `blake2_256`.
* Add a new `Get` type to the `Config` that constitutes the limit of the key. **Not** to the `Schedule`.
* Benchmark these functions with the maximum key size.
* I think a sensible default for this max key size could be 128byte.

This is needed for https://github.com/paritytech/ink/issues/1134

### PR DESCRIPTION
**Pull request by agryaznov**
Resolve #11029

cumulus companion: https://github.com/paritytech/cumulus/pull/1404

### CODE REVIEW

**Comment by agryaznov** [comment]
Still WIP but mostly done so could get some feedback already. 

Things left to do:
- [ ] add tests for newly added `_transparent` functions 
- [ ] add new _seal_ functions for storage ops which use `VarSizedKey`
- [ ] add benchmarks for these new _seal_ functions
- [ ] update ink! part

File: frame/contracts/src/exec.rs [hunk 1]
```diff
@@ -1085,24 +1119,48 @@ where
...
+			&frame.contract_info.get(&frame.account_id).trie_id,
+			key,
+			value,
+			Some(&mut frame.nested_storage),
+			take_old,
+		)
+	}
+
+	fn set_storage_transparent(
+		&mut self,
```
Comment (athei) [inline frame/contracts/src/exec.rs:1155]: Can have a default impl in the trait I think.

File: frame/contracts/src/wasm/runtime.rs [hunk 2]
```diff
@@ -705,22 +717,55 @@ where
 		if value_len > max_size {
 			return Err(Error::<E::T>::ValueTooLarge.into())
 		}
-		let mut key: StorageKey = [0; 32];
+
+		let mut key = vec![0; key_len.try_into().map_err(|_| Error::<E::T>::ValueTooLarge)?];
 		self.read_sandbox_memory_into_buf(key_ptr, &mut key)?;
```
Comment (athei) [inline frame/contracts/src/wasm/runtime.rs:722]: We are allocating memory based on user input here (`key_len` can be supplied by the contract). Needs to be capped by the maximum size.

```suggestion
                ensure!(key_len <= <E::T>::MaxStorageKeyLen, Error::<E::T>::DecodingFailed);
		let key = self.read_sandbox_memory(key_ptr, key_len)?
```

File: frame/contracts/src/exec.rs (hunk 1 above)
Comment (agryaznov) [reply frame/contracts/src/exec.rs:1155]: will be merged into single function using a generic argument, but we'll need to make tests not make it into an object, first

**Comment by agryaznov** [comment]
Progress so far: 

- [X] fixes from 1st review
- [X] add new seal_ functions for storage ops which use `VarSizedKey`
- [X] add benchmarks for these new seal_ funcs
- [X] add tests for newly added `_transparent` functions
- [ ] update wasm tests

File: frame/contracts/src/exec.rs [hunk 3]
```diff
@@ -140,19 +165,43 @@ pub trait Ext: sealing::Sealed {
 	///
 	/// Returns `None` if the `key` wasn't previously set by `set_storage` or
 	/// was deleted.
-	fn get_storage(&mut self, key: &StorageKey) -> Option<Vec<u8>>;
+	fn get_storage(&mut self, key: FixSizedKey) -> Option<Vec<u8>>;
```
Comment (athei) [inline frame/contracts/src/exec.rs:168]: Let's keep the `key` a reference. There is no reason to pass it by value,  right? Same for the other functions.

File: frame/contracts/src/lib.rs [hunk 4]
```diff
@@ -98,18 +98,18 @@ pub mod weights;
...
 	gas::GasMeter,
 	storage::{meter::Meter as StorageMeter, ContractInfo, DeletedContract, Storage},
 	wasm::{OwnerInfo, PrefabWasmModule},
 	weights::WeightInfo,
 };
+pub use crate::{
+	exec::{FixSizedKey, Frame, VarSizedKey},
+	pallet::*,
+	schedule::{HostFnWeights, InstructionWeights, Limits, Schedule},
+};
```
Comment (athei) [inline frame/contracts/src/lib.rs:112]: Why change the position? Let's keep the `pub` exports on top and not in-between the other imports.

File: frame/contracts/src/storage.rs [hunk 5]
```diff
@@ -124,16 +124,16 @@ where
 	///
 	/// The read is performed from the `trie_id` only. The `address` is not necessary. If the
 	/// contract doesn't store under the given `key` `None` is returned.
-	pub fn read(trie_id: &TrieId, key: &StorageKey) -> Option<Vec<u8>> {
-		child::get_raw(&child_trie_info(trie_id), &blake2_256(key))
+	pub fn read<K: StorageKey<T>>(trie_id: &TrieId, key: K) -> Option<Vec<u8>> {
```
[the remark this answers is not in the document]
Comment (agryaznov) [reply frame/contracts/src/storage.rs:127]: true thing, thanks

File: bin/node/runtime/src/lib.rs [hunk 6]
```diff
@@ -1937,7 +1938,7 @@ impl_runtime_apis! {
 			address: AccountId,
 			key: [u8; 32],
 		) -> pallet_contracts_primitives::GetStorageResult {
```
Comment (agryaznov) [inline bin/node/runtime/src/lib.rs:1940]: @athei I guess we should implement `get_storage_transparent()` for the RPC API here?

File: frame/contracts/src/wasm/mod.rs [hunk 7]
```diff
@@ -895,6 +920,84 @@ mod tests {
...
+		.unwrap();
+		ext.set_storage_transparent(
+			VarSizedKey::<Test>::try_from([2u8; 19].to_vec()).unwrap(),
+			Some(vec![]),
+			false,
+		)
+		.unwrap();
+
+		//value does not exist (wrong key length)
+		let input = (63, [1u8; 64]).encode();
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:979]: But you could just use a shorter input in this case?:
```rust
let input = [1u8; 63].encode();
```

File: frame/contracts/src/wasm/runtime.rs [hunk 8]
```diff
@@ -921,7 +994,7 @@ define_env!(Env, <E: Ext>,
 	// `ReturnCode::KeyNotFound`
 	[seal0] seal_get_storage(ctx, key_ptr: u32, out_ptr: u32, out_len_ptr: u32) -> ReturnCode => {
```
Comment (athei) [inline frame/contracts/src/wasm/runtime.rs:995]: Should probably be factored out to its own function on `Runtime` just as for `seal_set_storage`.

File: frame/contracts/src/wasm/runtime.rs [hunk 9]
```diff
@@ -945,7 +1052,7 @@ define_env!(Env, <E: Ext>,
 	// `SENTINEL` is returned as a sentinel value.
 	[seal0] seal_contains_storage(ctx, key_ptr: u32) -> u32 => {
```
Comment (athei) [inline frame/contracts/src/wasm/runtime.rs:1053]: Should probably be factored out to its own function on `Runtime` just as for `seal_set_storage`.

File: frame/contracts/src/benchmarking/mod.rs [hunk 10]
```diff
@@ -896,35 +902,43 @@ benchmarks! {
...
+	// it at a virgin key.
 	#[skip_meta]
 	seal_set_storage {
 		let r in 0 .. API_BENCHMARK_BATCHES;
-		let keys = (0 .. r * API_BENCHMARK_BATCH_SIZE)
-			.map(|n| T::Hashing::hash_of(&n).as_ref().to_vec())
-			.collect::<Vec<_>>();
-		let key_len = keys.get(0).map(|i| i.len() as u32).unwrap_or(0);
+		let max_key_len = T::MaxStorageKeyLen::get();
+		let keys = (0 .. r * API_BENCHMARK_BATCH_SIZE/2)
```
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/benchmarking/mod.rs:914]: I think you cannot just do that because the weight calculation in `schedule.rs` depends on that. Check the definition of the `cost_batched_args` macro there.

File: frame/contracts/src/wasm/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (agryaznov) [reply frame/contracts/src/wasm/mod.rs:979]: this would work only for the case when the passed `key_len` is wrong, as
```rust
let input = [1u8; 63].encode(); // key_len = "\49\02" (585u32), key = [1u8;59]: wrong key_len (585 != 59)
```
while other cases (correct `key_len`) would still need it like this
```rust
let input = (64, [1u8; 63]).encode();
```
so I decided to keep it uniformly along all cases, for better readability.

File: frame/contracts/src/wasm/runtime.rs [hunk 11]
```diff
@@ -705,26 +729,127 @@ where
...
+					&FixSizedKey::try_from(key).map_err(|_| Error::<E::T>::DecodingFailed)?,
+				) {
+					self.adjust_gas(charged, RuntimeCosts::GetStorage(value.len() as u32));
+					self.write_sandbox_output(
+						out_ptr,
+						out_len_ptr,
+						&value,
+						false,
+						already_charged,
+					)?;
```
Comment (athei) [inline frame/contracts/src/wasm/runtime.rs:795]: This part seems to be the same for both branches. Can't it be factored out as in `clear_storage`?

File: frame/contracts/src/wasm/runtime.rs [hunk 12]
```diff
@@ -705,26 +729,127 @@ where
...
+			KeyType::Fix => {
+				if let Some(len) = self.ext.get_storage_size(
+					&FixSizedKey::try_from(key).map_err(|_| Error::<E::T>::DecodingFailed)?,
+				) {
+					self.adjust_gas(charged, RuntimeCosts::ContainsStorage(len));
+					Ok(len)
+				} else {
+					self.adjust_gas(charged, RuntimeCosts::ContainsStorage(0));
+					Ok(SENTINEL)
+				}
```
Comment (athei) [inline frame/contracts/src/wasm/runtime.rs:836]: This part seems to be the same for both branches. Can't it be factored out as in `clear_storage`?

File: frame/contracts/src/wasm/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:979]: Why? You can use the length written by `seal_input` in both cases. In one case you write a too short of an input and in the other case you don't. You don't need to separately signal the size of the buffer. `seal_input` communicates that.

File: frame/contracts/src/lib.rs [hunk 13]
```diff
@@ -942,11 +946,14 @@ where
...
 	/// Query storage of a specified contract under a specified key.
-	pub fn get_storage(address: T::AccountId, key: [u8; 32]) -> GetStorageResult {
+	pub fn get_storage(address: T::AccountId, key: Vec<u8>) -> GetStorageResult {
 		let contract_info =
 			ContractInfoOf::<T>::get(&address).ok_or(ContractAccessError::DoesntExist)?;
 
-		let maybe_value = Storage::<T>::read(&contract_info.trie_id, &key);
+		let maybe_value = Storage::<T>::read(
+			&contract_info.trie_id,
+			&VarSizedKey::<T>::try_from(key).map_err(|_| ContractAccessError::DoesntExist)?,
```
Comment (athei) [inline frame/contracts/src/lib.rs:955]: This error doesn't exactly fit what is happening here. You might want to add another one. Even better: Already use a `VarSizedKey` as an argument here to push the error further out.

File: bin/node/runtime/src/lib.rs (hunk 6 above)
Comment (agryaznov) [reply bin/node/runtime/src/lib.rs:1947]: So making `Pallet::get_storage` accept `VarSizedKey` leads to necessity of adding this type to `node-runtime` and `node-rpc` which doesn't seem right way to do to me. After some attempts (see todays commits) I decided not to go in that direction, revert it as it was done at first with `Vec<u8>` and add a new `ContractAccessError` error

File: frame/contracts/src/benchmarking/mod.rs [hunk 14]
```diff
@@ -896,35 +902,43 @@ benchmarks! {
...
+					offset: 32,
 					value: key_bytes,
 				},
 			],
 			call_body: Some(body::repeated_dyn(r * API_BENCHMARK_BATCH_SIZE, vec![
-				Counter(0, key_len as u32), // key_ptr
-				Regular(Instruction::I32Const(0)), // value_ptr
-				Regular(Instruction::I32Const(0)), // value_len
+				Counter(32, max_key_len as u32), // key_ptr
+				Regular(Instruction::I32Const(0)), // key_len
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:939]: This seems wrong. Shouldn't `key_len` be `max_key_len`?

File: frame/contracts/src/benchmarking/mod.rs [hunk 15]
```diff
@@ -947,29 +961,35 @@ benchmarks! {
...
+				DataSegment {
+					offset: 32,
 					value: key_bytes,
 				},
 			],
 			call_body: Some(body::repeated_dyn(API_BENCHMARK_BATCH_SIZE, vec![
-				Counter(0, key_len as u32), // key_ptr
-				Regular(Instruction::I32Const(0)), // value_ptr
+				Counter(32, max_key_len as u32), // key_ptr
+				Regular(Instruction::I32Const(0)), // key_len
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:991]: Same here. `key_len` seems wrong.

File: frame/contracts/src/benchmarking/mod.rs [hunk 16]
```diff
@@ -896,35 +902,43 @@ benchmarks! {
...
 				name: "seal_set_storage",
-				params: vec![ValueType::I32, ValueType::I32, ValueType::I32],
+				params: vec![ValueType::I32, ValueType::I32, ValueType::I32, ValueType::I32],
 				return_type: Some(ValueType::I32),
 			}],
 			data_segments: vec![
 				DataSegment {
 					offset: 0,
+					value: max_key_len.to_le_bytes().to_vec(),
+				},
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:931]: `key_len` is not a pointer.

File: frame/contracts/src/benchmarking/mod.rs [hunk 17]
```diff
@@ -947,29 +961,35 @@ benchmarks! {
...
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:983]: `key_len` is not a pointer.

File: frame/contracts/src/wasm/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:979]: But if you pass `[1u8; 63]`  instead of `(63, [1u8; 64]).encode()` wouldn't you be testing the same thing? `seal_input` would communicate `63` as `key_len`.

File: frame/contracts/src/wasm/mod.rs [hunk 18]
```diff
@@ -425,27 +426,51 @@ mod tests {
...
+			let entry = self.storage.entry(key.to_vec());
+			let result = match (entry, take_old) {
+				(Entry::Vacant(_), _) => WriteOutcome::New,
+				(Entry::Occupied(entry), false) =>
+					WriteOutcome::Overwritten(entry.remove().len() as u32),
+				(Entry::Occupied(entry), true) => WriteOutcome::Taken(entry.remove()),
+			};
+			if let Some(value) = value {
+				self.storage.insert(key.to_vec(), value);
 			}
```
Comment (athei) [inline frame/contracts/src/wasm/mod.rs:474]: We should de duplicate this code while we can't have these functions itself generic: Create a generic function in this module which you call from both of these function.

File: frame/contracts/src/lib.rs [hunk 19]
```diff
@@ -129,6 +124,12 @@ use sp_core::{crypto::UncheckedFrom, Bytes};
 use sp_runtime::traits::{Convert, Hash, Saturating, StaticLookup};
 use sp_std::{fmt::Debug, marker::PhantomData, prelude::*};
 
+pub use crate::{
+	exec::{FixSizedKey, Frame, VarSizedKey},
```
[the remark this answers is not in the document]
Comment (agryaznov) [reply frame/contracts/src/lib.rs:128]: e.g. `node-executor` basic test uses it [here](https://github.com/paritytech/substrate/blob/ag-transparent-hashing/bin/node/executor/tests/basic.rs#L769)

[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/lib.rs:128]: But that is not necessary? It would be best if you didn't introduce new users of deprecated types. Just use a `VarSizedKey` there. Then the fixed key can be private. Also, we should export `VarSizedKey` just as `StorageKey`. Let's hide the fixed-size key or the fact that there are multiple key types as well as we can. It is just a detail for pre-existing contracts.

File: frame/contracts/src/wasm/mod.rs (hunk 18 above)
Comment (agryaznov) [reply frame/contracts/src/wasm/mod.rs:474]: What signature do you mean for that function? if its like this 
```rust
	fn set_storage_generic<K: StorageKey<T>>(
		&mut self,
		key: &K,
		value: Option<Vec<u8>>,
		take_old: bool,
	) -> Result<WriteOutcome, DispatchError> {

```
then this won't compile with 
```
`exec::Ext` cannot be made into an object
```

This could be made like this
```rust
	fn set_storage_generic(
		&mut self,
		key: Vec<u8>,
		value: Option<Vec<u8>>,
		take_old: bool,
	) -> Result<WriteOutcome, DispatchError> {
```

but once `storage::write()` function [accepts](https://github.com/paritytech/substrate/blob/ag-transparent-hashing/frame/contracts/src/storage.rs#L146) generic parameter, this won't make much sense either

File: frame/contracts/src/wasm/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (agryaznov) [reply frame/contracts/src/wasm/mod.rs:979]: If I pass `[1u8; 63]` I would test versus `[1u8; 63]` `key` and `63` `key_len`. 
I pass `(63, [1u8; 64])` to test versus `[1u8; 64]` `key` and `63` `key_len`.

[the remark this answers is not in the document]
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:979]: Okay, I see why this tests more things. Like some bug where the key length is not properly honored.

File: frame/contracts/src/benchmarking/mod.rs [hunk 20]
```diff
@@ -896,35 +902,40 @@ benchmarks! {
...
+					value: keys_bytes,
 				},
 			],
 			call_body: Some(body::repeated_dyn(r * API_BENCHMARK_BATCH_SIZE, vec![
-				Counter(0, key_len as u32), // key_ptr
+				Counter(0, max_key_len as u32), // key_ptr
+				Regular(Instruction::I32Const(max_key_len as i32)), // key_len
 				Regular(Instruction::I32Const(0)), // value_ptr
-				Regular(Instruction::I32Const(0)), // value_len
+				Regular(Instruction::I32Const(value_len as i32)), // value_len
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:938]: This test should test the overhead of writing a zero-length value. Why the change?

File: frame/contracts/src/wasm/mod.rs (hunk 18 above)
Comment (athei) [reply frame/contracts/src/wasm/mod.rs:474]: I didn't realize that this is only the `MockExt`. This is fine then.

File: frame/contracts/src/benchmarking/mod.rs [hunk 21]
```diff
@@ -968,9 +979,10 @@ benchmarks! {
 				},
 			],
 			call_body: Some(body::repeated_dyn(API_BENCHMARK_BATCH_SIZE, vec![
-				Counter(0, key_len as u32), // key_ptr
+				Counter(0, max_key_len as u32), // key_ptr
+				Regular(Instruction::I32Const(max_key_len as i32)), // key_len
 				Regular(Instruction::I32Const(0)), // value_ptr
-				Regular(Instruction::I32Const((n * 1024) as i32)), // value_len
+				Regular(Instruction::I32Const((n * 2048) as i32)), // value_len = max payload_len
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:985]: IMHO the comment is wrong. It **varies** between `0` and `pax_payload_len / 2`.

File: frame/contracts/src/benchmarking/mod.rs [hunk 22]
```diff
@@ -1119,8 +1138,8 @@ benchmarks! {
 		for key in keys {
 			Storage::<T>::write(
 				&info.trie_id,
-				key.as_slice().try_into().map_err(|e| "Key has wrong length")?,
-				Some(vec![42u8; (n * 1024) as usize]),
+				&VarSizedKey::<T>::try_from(key).map_err(|e| "Key has wrong length")?,
+				Some(vec![42u8; (n * 2048) as usize]), // value_len = max payload_len
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:1142]: IMHO the comment is wrong. It varies between 0 and pax_payload_len / 2.

File: frame/contracts/src/benchmarking/mod.rs [hunk 23]
```diff
@@ -1027,8 +1041,8 @@ benchmarks! {
...
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:1045]: IMHO the comment is wrong. It varies between 0 and pax_payload_len / 2.

File: frame/contracts/src/benchmarking/mod.rs [hunk 24]
```diff
@@ -1223,8 +1246,8 @@ benchmarks! {
...
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:1250]: IMHO the comment is wrong. It varies between 0 and pax_payload_len / 2.

File: frame/contracts/src/benchmarking/mod.rs [hunk 25]
```diff
@@ -1315,8 +1342,8 @@ benchmarks! {
...
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:1346]: IMHO the comment is wrong. It varies between 0 and pax_payload_len / 2.

File: frame/contracts/src/benchmarking/mod.rs [hunk 26]
```diff
@@ -1419,8 +1450,8 @@ benchmarks! {
...
```
Comment (athei) [inline frame/contracts/src/benchmarking/mod.rs:1454]: IMHO the comment is wrong. It varies between 0 and pax_payload_len / 2.

File: bin/node/executor/tests/basic.rs [hunk 27]
```diff
@@ -766,15 +766,15 @@ fn deploying_wasm_contract_should_work() {
 		// It does not matter that the storage item itself does not exist.
 		assert!(&pallet_contracts::Pallet::<Runtime>::get_storage(
 			addr,
-			pallet_contracts::StorageKey::default().to_vec()
+			pallet_contracts::StorageKey<Runtime>::default().to_vec()
```
Comment (KiChjang) [inline bin/node/executor/tests/basic.rs]:
```suggestion
			pallet_contracts::StorageKey::<Runtime>::default().to_vec()
```

**Review by athei — APPROVED** [review]

File: frame/contracts/rpc/src/lib.rs [hunk 28]
```diff
@@ -74,6 +74,12 @@ impl From<ContractAccessError> for JsonRpseeError {
 				None::<()>,
 			))
 			.into(),
+			KeyDecodingFailed => CallError::Custom(ErrorObject::owned(
+				CONTRACT_DOESNT_EXIST,
```
Comment (athei) [inline frame/contracts/rpc/src/lib.rs:78]: Please define a new constant here with another error code.

**Comment by athei** [comment]
We need to re-run the bot and have a look at the weights. This is the best way to catch bugs in benchmarks: Checking if the weights are plausible.

File: frame/contracts/src/wasm/mod.rs [hunk 29]
```diff
@@ -425,27 +426,51 @@ mod tests {
...
+		) -> Result<WriteOutcome, DispatchError> {
+			let entry = self.storage.entry(key.to_vec());
+			let result = match (entry, take_old) {
+				(Entry::Vacant(_), _) => WriteOutcome::New,
+				(Entry::Occupied(entry), false) =>
+					WriteOutcome::Overwritten(entry.remove().len() as u32),
+				(Entry::Occupied(entry), true) => WriteOutcome::Taken(entry.remove()),
+			};
+			if let Some(value) = value {
+				self.storage.insert(key.to_vec(), value);
```
Comment (ascjones) [inline frame/contracts/src/wasm/mod.rs:473]: Minor optimization could call `key.to_vec()` just once at the top

**Review by ascjones — APPROVED** [review]
LGTM

**Comment by athei** [comment]
I noticed that all changed `*_per_kb` benchmarks suddenly include a read/write component per `n`. We don't want this. I think in those benchmarks we need to whitelist the generated keys using: [add_to_whitelist](https://paritytech.github.io/substrate/master/frame_benchmarking/benchmarking/fn.add_to_whitelist.html).

**Review by rcny — APPROVED** [review]

**Review by TriplEight — APPROVED** [review]

**Comment by agryaznov** [comment]
bot merge

**Comment by agryaznov** [comment]
bot merge
